### PR TITLE
Create project registry

### PR DESCRIPTION
**Pull request by mzxyz**
## Changes

- Create `ProjectRegistry` contract
- Support NFT features for ProjectRegistry
- Remove unused functions

### CODE REVIEW

File: contracts/ProjectRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+ * This contract tracks all projects and their deployments. At the beginning of the network,
+ * we will start with the restrict mode which only allow permissioned account to create and update project.
+ * Indexers are able to start and stop indexing with a specific deployment from this conttact. Also Indexers can update and report
+ * their indexing status from this contarct.
+ */
+contract ProjectRegistry is Initializable, OwnableUpgradeable, ERC721Upgradeable, IProjectRegistry {
+    /// @notice project information
+    struct ProjectInfo {
+        uint256 projectId;
+        address owner;
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]: we can use nft's owner

File: contracts/ProjectRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        bytes32 latestVersion;
+        bytes32 latestDeploymentId;
+        bytes32 metadata;
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol:31]: we can reuse nft's metadata

Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]: can we use `erc-721's metadata extension` for metadata?

**Comment by ianhe8x** [comment]
We can consider to support `The enumeration extension is OPTIONAL for ERC-721 smart contracts (see “caveats”, below). This allows your contract to publish its full list of NFTs and make them discoverable.`
https://eips.ethereum.org/EIPS/eip-721

File: contracts/ProjectRegistry.sol (hunk 1 above)
Comment (mzxyz) [reply contracts/ProjectRegistry.sol]: Will do

File: contracts/ProjectRegistry.sol (hunk 2 above)
Comment (mzxyz) [reply contracts/ProjectRegistry.sol]: We can use metadata from ERC-721

File: contracts/ProjectRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+ * @notice ### Overview
+ * This contract tracks all projects and their deployments. At the beginning of the network,
+ * we will start with the restrict mode which only allow permissioned account to create and update project.
+ * Indexers are able to start and stop indexing with a specific deployment from this conttact. Also Indexers can update and report
+ * their indexing status from this contarct.
+ */
+contract ProjectRegistry is Initializable, OwnableUpgradeable, ERC721Upgradeable, IProjectRegistry {
+    /// @notice project information
+    struct ProjectInfo {
+        uint256 projectId;
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]: can deprecate it now

File: contracts/ProjectRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    mapping(address => bool) public creatorWhitelist;
+
+    /// @notice deployment id -> indexer -> IndexingStatus
+    mapping(bytes32 => mapping(address => IndexingStatus)) public deploymentStatusByIndexer;
+
+    /// @notice indexer -> deployment numbers
+    mapping(address => uint256) public numberOfIndexingDeployments;
+
+    /// @notice is the id a deployment
+    mapping(bytes32 => bool) private deploymentIds;
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]: allow deployment to project lookup

File: contracts/ProjectRegistry.sol (hunk 2 above)
Comment (ianhe8x) [reply contracts/ProjectRegistry.sol]:
```
struct ProjectInfo {
  bytes32 latestDeploymentId;
  ProjectType projectType;
}
struct DeploymentInfo {
  uint256 projectId
  bytes32 metadata // aka version
}
mapping(bytes32 => DeploymentInfo) private deploymentIds;
```

**Comment by ianhe8x** [comment]
@mzxyz have you added the change to merge `indexing` with `ready`?
also the change to allow indexer stop receiving new agreement

File: contracts/ProjectRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    function removeCreator(address creator) external onlyOwner {
+        creatorWhitelist[creator] = false;
+    }
+
+    function setBaseURI(string memory baseURI_) external onlyOwner() {
+        baseURI = baseURI_;
+    }
+
+    function _baseURI() internal view virtual override returns (string memory) {
+        return baseURI;
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]:
```suggestion
        return "ipfs://";
```
there's no need to change base after the contract creation.

File: contracts/ProjectRegistry.sol [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        projectInfos[projectId].latestDeploymentId = deploymentId;
+        deploymentInfos[deploymentId] = DeploymentInfo(projectId, metadata);
+
+        emit UpdateProjectDeployment(msg.sender, projectId, deploymentId, metadata);
+    }
+
+    /**
+     * @notice Indexer update its indexing status to ready with a specific deploymentId
+     */
+    function updateIndexingStatusToReady(bytes32 deploymentId) external onlyIndexer {
```
Comment (ianhe8x) [inline contracts/ProjectRegistry.sol]: replace `Indexing` to more general name to support rpc

**Comment by mzxyz** [comment]
@ianhe8x In Solidity, when using super with multiple inheritance, the function calls are determined by the C3 linearization of the inheritance graph (or the C3 superclass linearization). In practice, it means that when you use super, the function that gets called is the next one in line based on the linearized order of inheritance, starting from the current contract and moving right to left through its base contracts. Once it finds and executes a matching function in this order, it stops.

https://docs.soliditylang.org/en/develop/contracts.html#multiple-inheritance-and-linearization

**Pull request merged**